junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/ClassBasedTestDescriptor.java

```diff
@@ -295,27 +300,35 @@ private TestInstances instantiateAndPostProcessTestInstance(JupiterEngineExecuti
 	}
 
 	protected abstract TestInstances instantiateTestClass(JupiterEngineExecutionContext parentExecutionContext,
-			ExtensionRegistry registry, JupiterEngineExecutionContext context);
+			ExtensionContext ourExtensionContext, ExtensionRegistry registry, JupiterEngineExecutionContext context);
 
 	protected TestInstances instantiateTestClass(Optional<TestInstances> outerInstances, ExtensionRegistry registry,
-			ExtensionContext extensionContext) {
+			ExtensionContext extensionContext, ExtensionContext ourExtensionContext) {
 
 		Optional<Object> outerInstance = outerInstances.map(TestInstances::getInnermostInstance);
 		invokeTestInstancePreConstructCallbacks(new DefaultTestInstanceFactoryContext(this.testClass, outerInstance),
-			registry, extensionContext);
+			registry, extensionContext, ourExtensionContext);
 		Object instance = this.testInstanceFactory != null //
-				? invokeTestInstanceFactory(outerInstance, extensionContext) //
-				: invokeTestClassConstructor(outerInstance, registry, extensionContext);
+				? invokeTestInstanceFactory(outerInstance, extensionContext, ourExtensionContext) //
+				: invokeTestClassConstructor(outerInstance, registry, extensionContext, ourExtensionContext);
 		return outerInstances.map(instances -> DefaultTestInstances.of(instances, instance)).orElse(
 			DefaultTestInstances.of(instance));
 	}
 
-	private Object invokeTestInstanceFactory(Optional<Object> outerInstance, ExtensionContext extensionContext) {
+	private Object invokeTestInstanceFactory(Optional<Object> outerInstance, ExtensionContext extensionContext,
+			ExtensionContext ourExtensionContext) {
 		Object instance;
 
 		try {
-			instance = this.testInstanceFactory.createTestInstance(
-				new DefaultTestInstanceFactoryContext(this.testClass, outerInstance), extensionContext);
+			if (AnnotationUtils.isAnnotated(this.testInstanceFactory.getClass(),
+				EnableTestScopedConstructorContext.class)) {
```

Comment:
I am a little bit concerned about performance with all these calls to `AnnotationUtils.isAnnotated`. Do you think this could be an issue?

Reply:
It does seem unnecessary to check it multiple times (in different places). I'll take care of that in a follow-up PR.

Reply:
Ok, if you have an idea how to solve this, then it is good. My best idea would be to add a cache to the extension registry.

Just want to note that using a marker interface instead of an annotation would also alleviate the problem, as `instanceof` is rather fast as far as I know. However, I agree that a marker interface is probably not the best solution on extensions, which may already be rather heavy on interfaces.

Not sure if that is the right place to discuss it, but I also just had another idea. We could introduce a common method with a default implementation to all the affected extension interfaces.

```java
default boolean supportsTestScopedConstructorContext() {
    // or isMethodLevelExtensionContextAware()
    return false;
}
```

Not entirely sure if I would like such a method. It seems like a very good solution from the technical perspective, but it might feel a bit clunky.

Reply:
> Ok, if you have an idea how to solve this, then it is good. My best idea would be to add a cache to the extension registry, which is probably fine.

Yes, something along those lines. But I'd like to get feel for it myself before I can make a concrete recommendation. In any case, I think this can be addressed separately.

> Just want to note that using a marker interface instead of an annotation would also alleviate the problem, as `instanceof` is rather fast as far as I know. However, I agree that a marker interface is probably not the best solution on extensions, which may already be rather heavy on interfaces.
> 
> Not sure if that is the right place to discuss it, but I also just had another idea. We could introduce a common method with a default implementation to all the affected extension interfaces.

Thanks for sharing your ideas! I think both would be equivalent but I still think the annotation is the best option for a "temporary" solution like this because it doesn't muddle the extension interfaces even if it's a little more involved implementation-wise.